Improved constraint handling, part 3

nyalldawson opened the pull request:
Improves handling of defaults (inc provider default clauses, literal defaults, and qgis expression defaults) and automatically handling unique value constraints on layers.

This adds a new method QgsVectorLayerUtils::createFeature which returns a new feature which includes all relevant defaults. Any fields with unique value constraints will be guaranteed to have a value
which is unique for the field. It's heavily unit tested and designed to replace all the duplicate, inconsistent code for handling provider default clauses currently scattered around the code base.

All (identified) map tools which create new features have been updated to use this to gain automatic handling of provider defaults, qgis default expressions, and provider/qgis side unique field constraints.

Sponsored by Canton of Zug and the QGEP project

nyalldawson committed:
[FEATURE] Add method to get list of unique strings matching
a substring from a vector data provider

Base implementation iterates through all features, but
providers can override with optimised versions. Native
implementations for postgres, spatialite and OGR included.

nyalldawson committed:
Add QgsVectorLayer::uniqueStringsMatching() to retrieve
a list of unique strings matching a substring for a field

nyalldawson committed:
New method QgsVectorLayerUtils::createUniqueValue

Returns a new unique attribute value for a layer. For numeric
fields this is the max attribute value + 1, for strings we
attempt to create a unique value by append _1, _2, etc to either
a specified 'seed' value, or failing that to the end of the
value of that field from the first feature in the layer.

nyalldawson committed:
Fix clearing constraints

nyalldawson committed:
[FEATURE] Improve handling of defaults (inc provider default clauses,
literal defaults, and qgis expression defaults) and automatically
handle unique value constraints on layers

Add a new method QgsVectorLayerUtils::createFeature which returns
a new feature which includes all relevant defaults. Any fields
with unique value constraints will be guaranteed to have a value
which is unique for the field.

Currently only in use by the split feature tool.

Sponsored by Canton of Zug and the QGEP project

nyalldawson commented:
Video demo at: https://www.youtube.com/watch?v=gG0Yg5JRDP8&feature=youtu.be

nyalldawson committed:
Make sure primary keys are always marked with a unique, not null
constraint for all providers

nyalldawson committed:
Port remaining map tools, paste features, offline editing
to use new constraint/default clause handling methods

nyalldawson commented:
@m-kuhn should be all ready to go now. Do you mind taking a look?

In regards to the new QgsVectorLayer::uniqueStringsMatching() methods, this has been implemented both for this work (helping to create a new unique string value for string fields with a unique constraint) AND also some other work I've had sitting around which adds a field value autocompleter to search widget wrappers. So that's why it's implemented in this way, with the feedback option, etc.

m-kuhn commented:
I'll test it today

nyalldawson commented:
> One thing that confuses me, when a (provider side) unique constraint is set on a field and the provider also provides default values, is still an autogenerated one used?

In this case the provider default clause takes precedence. Can you confirm what edit widget you have? I noticed an issue where numeric field with a provider side default get auto converted to range widgets and the default is lost. Maybe it's this same issue.

m-kuhn commented:
![unique](https://cloud.githubusercontent.com/assets/588407/20209337/2fcb1e9e-a7f4-11e6-8449-759dd3e4d194.gif)

nyalldawson commented:
Mind sharing the table definition?

m-kuhn commented:
https://github.com/QGEP/data/archive/demodata.zip

nyalldawson committed:
Prioritise provider default literals over reused values when
creating a new feature

nyalldawson committed:
Avoid double-evaluating postgres default values

nyalldawson committed:
Don't use auto generated widgets for fields with provider default
value clause

Using auto widgets may cause the default value clause to be mangled,
eg by converting it to a number

nyalldawson commented:
@m-kuhn Thanks, can you try the latest version? Should be fixed now.

m-kuhn commented:
Too cool, it even works with default values on views!
